File: modules/16-context-engineering/ts/03-memory-compaction.ts
/**
 * Task 3 — Conversation memory / compaction 🟡
 *
 * What this teaches:
 *   - In a long chat, the context window fills up with old turns. The model sees the
 *     full history on every request — you are responsible for keeping it within budget.
 *   - Sliding window (keep the last K turns) is simple but loses early context.
 *   - Running summary (summarise old turns when the budget is exceeded) is more powerful:
 *     the model retains the gist without the full verbatim transcript.
 *   - A hybrid (summary + recent turns) is the most robust production approach.
 *
 * How to run:
 *   pnpm tsx modules/16-context-engineering/ts/03-memory-compaction.ts
 */

import "dotenv/config";
import { getProvider, ChatMessage, ChatOptions } from "@learn-ai/llm-core";

// ---------------------------------------------------------------------------
// Rough token counter (replace with tiktoken from Task 1 for precision).
// ---------------------------------------------------------------------------
function countTokens(text: string): number {
  return Math.ceil(text.split(/\s+/).length * 1.3);
}

function countHistoryTokens(messages: ChatMessage[]): number {
  return messages.reduce((sum, m) => sum + countTokens(m.content), 0);
}

// Budget is intentionally small to trigger compaction during the demo.
const TOKEN_BUDGET = 800;

// ---------------------------------------------------------------------------
// Simulated conversation — 15 turns covering distinct topics.
// The early turns (movie recommendation) will eventually be compacted.
// ---------------------------------------------------------------------------
const CONVERSATION_SCRIPT: Array<[string, string]> = [
  ["user",      "What's a good sci-fi movie to watch tonight?"],
  ["assistant", "I'd recommend 'Arrival' (2016). It's a cerebral sci-fi film about a linguist who helps decode an alien language. Thoughtful, slow-paced, and emotionally resonant."],
  ["user",      "Sounds great! Is it suitable for kids?"],
  ["assistant", "It's rated PG-13. The themes are abstract rather than violent, but younger children might find the pacing slow. Fine for teenagers and adults."],
  ["user",      "OK. Now let's talk about TypeScript. How do I read a file?"],
  ["assistant", "Use `fs.readFileSync('file.txt', 'utf-8')` for synchronous reads, or `fs.promises.readFile('file.txt', 'utf-8')` for async. The async approach is preferred in modern Node.js."],
  ["user",      "What about writing to a file?"],
  ["assistant", "Use `fs.writeFileSync('output.txt', content)` or `await fs.promises.writeFile('output.txt', content)`. For appending, pass `{ flag: 'a' }` as the third argument."],
  ["user",      "How does the Node.js event loop work?"],
  ["assistant", "The event loop processes tasks in phases: timers (setTimeout/setInterval), I/O callbacks, idle/prepare, poll (new I/O), check (setImmediate), and close callbacks. Each phase has a FIFO queue; the loop continues until all queues are empty."],
  ["user",      "Got it. Can you explain Promise.all vs Promise.allSettled?"],
  ["assistant", "Promise.all resolves when ALL promises resolve, or rejects as soon as ONE rejects. Promise.allSettled always resolves with an array of results (fulfilled or rejected) — it never rejects."],
  ["user",      "Thanks! Switching topics — what is the capital of Japan?"],
  ["assistant", "Tokyo."],
  ["user",      "And Australia?"],
  ["assistant", "Canberra."],
  ["user",      "One more — what was the movie you recommended earlier? Do you remember?"],
  ["user",      "It was at the very start of our chat."],
];

// ---------------------------------------------------------------------------
// TODO 1: Implement summariseTurns.
//         Given an array of ChatMessage objects, ask the LLM to compress them.
//         - Flatten `turns` into one transcript string ("<role>: <content>" per line).
//         - Build a prompt asking the model to summarise the exchange in 2–3 sentences,
//           keeping key facts / decisions / topics and dropping conversational filler,
//           with the transcript appended.
//         - Send it as a single user message via `llm.chat([...], { temperature: 0 })`
//           so the summary is deterministic.
//         - Return a ChatMessage with role "system" whose content begins with
//           "Summary of earlier conversation: " followed by the model's text (that
//           prefix is what maybeCompact and the model rely on to spot the summary).
// ---------------------------------------------------------------------------
async function summariseTurns(
  turns: ChatMessage[],
  llm: ReturnType<typeof getProvider>,
): Promise<ChatMessage> {
  throw new Error("TODO: implement summariseTurns");
}

// ---------------------------------------------------------------------------
// TODO 2: Implement maybeCompact.
//         1. Count the total tokens in `messages`.
//         2. If count > budget:
//            a. Separate system messages from non-system messages.
//            b. Find the oldest non-system turns to summarise (enough to go below budget).
//            c. Call summariseTurns on those oldest turns.
//            d. Return: [systemMessages, summaryMessage, ...remainingTurns].
//         3. If within budget, return messages unchanged.
//         4. Log when compaction fires.
// ---------------------------------------------------------------------------
async function maybeCompact(
  messages: ChatMessage[],
  budget: number,
  llm: ReturnType<typeof getProvider>,
): Promise<ChatMessage[]> {
  // - Use countHistoryTokens(messages); if it is within `budget`, return messages
  //   unchanged (fast path — no LLM call).
  // - Otherwise log a line noting compaction fired (so you can watch it in the demo),
  //   then split messages into system messages and conversation (non-system) turns.
  // - Choose how many of the OLDEST conversation turns to compress — summarising
  //   roughly the first half (at least one) is a reasonable split. Call summariseTurns
  //   on that oldest slice and keep the rest verbatim.
  // - Rebuild in order: system messages, then the single summary message, then the
  //   kept recent turns — and return that new array.
  throw new Error("TODO: implement maybeCompact");
}

async function main() {
  console.log("=== Task 3: Conversation Memory / Compaction ===\n");
  console.log(`Token budget: ${TOKEN_BUDGET} tokens\n`);

  const llm = getProvider();
  console.log(`Provider: ${llm.name} / ${llm.chatModel}\n`);

  // -------------------------------------------------------------------------
  // TODO 3: Simulate the conversation.
  //         For each turn in CONVERSATION_SCRIPT:
  //           1. Append the scripted message to history.
  //           2. If the turn is a "user" message, call maybeCompact, then
  //              call llm.chat(history) to get the assistant's reply.
  //           3. Append the assistant's reply to history.
  //           4. Print the current token count after each turn.
  // -------------------------------------------------------------------------

  let history: ChatMessage[] = [
    { role: "system", content: "You are a helpful, knowledgeable assistant." },
  ];

  for (const [role, content] of CONVERSATION_SCRIPT) {
    history.push({ role: role as "user" | "assistant", content });
    const tokens = countHistoryTokens(history);
    console.log(`[${role.padStart(9)}] tokens=${String(tokens).padStart(4)} | ${content.slice(0, 60)}...`);

    if (role === "user") {
      try {
        history = await maybeCompact(history, TOKEN_BUDGET, llm);
        const result = await llm.chat(history);
        history.push({ role: "assistant", content: result.text });
        console.log(`[assistant] ${result.text.slice(0, 80)}...`);
      } catch (e: unknown) {
        const msg = e instanceof Error ? e.message : String(e);
        if (msg.startsWith("TODO")) {
          console.log("  [TODO: implement maybeCompact and summariseTurns]");
        } else {
          console.log(`  ERROR: ${msg}`);
        }
        history.push({ role: "assistant", content: "(not implemented)" });
      }
    }
  }

  console.log();
  console.log("Observation:");
  console.log("  When compaction fires, observe whether the model still answers questions");
  console.log("  about early turns (e.g. the movie recommendation) via the summary.");
}

main().catch(console.error);

File: modules/16-context-engineering/py/03_memory_compaction.py
"""
Task 3 — Conversation memory / compaction 🟡

What this teaches:
  - In a long chat, the context window fills up with old turns. The model sees the
    full history on every request — you are responsible for keeping it within budget.
  - Sliding window (keep the last K turns) is simple but loses early context entirely.
  - Running summary (summarise old turns when the budget is exceeded) is more powerful:
    the model retains the gist of the conversation without the verbatim transcript.
  - A hybrid (summary + recent turns) is the most robust production approach.

How to run:
  uv run python modules/16-context-engineering/py/03_memory_compaction.py
"""

from __future__ import annotations

from llm_core import get_provider, ChatMessage, ChatOptions

# ---------------------------------------------------------------------------
# Rough token counter (replace with tiktoken in Task 1 if you want precision).
# ---------------------------------------------------------------------------
def count_tokens(text: str) -> int:
    return int(len(text.split()) * 1.3)


def count_history_tokens(messages: list[ChatMessage]) -> int:
    return sum(count_tokens(m.content) for m in messages)


# Budget is intentionally small to trigger compaction during the demo.
TOKEN_BUDGET = 800

# ---------------------------------------------------------------------------
# Simulated conversation — 15 turns covering several distinct topics.
# The early turns (the movie recommendation) will eventually be compacted.
# ---------------------------------------------------------------------------
CONVERSATION_SCRIPT = [
    ("user",      "What's a good sci-fi movie to watch tonight?"),
    ("assistant", "I'd recommend 'Arrival' (2016). It's a cerebral sci-fi film about a linguist who helps decode an alien language. Thoughtful, slow-paced, and emotionally resonant."),
    ("user",      "Sounds great! Is it suitable for kids?"),
    ("assistant", "It's rated PG-13. The themes are abstract rather than violent, but younger children might find the pacing slow. I'd say fine for teenagers and adults."),
    ("user",      "OK. Now let's talk about Python. How do I read a file line by line?"),
    ("assistant", "Use `with open('file.txt') as f: for line in f: print(line.strip())`. The `with` statement ensures the file is closed even if an exception occurs."),
    ("user",      "What about writing to a file?"),
    ("assistant", "Use `with open('output.txt', 'w') as f: f.write('hello\\n')`. For appending instead of overwriting, use mode `'a'` instead of `'w'`."),
    ("user",      "How does Python's GIL affect multi-threading?"),
    ("assistant", "The Global Interpreter Lock (GIL) prevents true parallel execution of Python threads for CPU-bound work. Use `multiprocessing` for CPU parallelism; threading is fine for I/O-bound work where threads spend most time waiting."),
    ("user",      "Got it. Can you explain async/await?"),
    ("assistant", "Async/await is Python's syntax for cooperative concurrency. An `async def` function is a coroutine. `await` suspends the coroutine until a result is ready, yielding control to the event loop. It's ideal for I/O-bound workloads without threads."),
    ("user",      "Thanks! Switching topics — what's the capital of France?"),
    ("assistant", "Paris."),
    ("user",      "And Germany?"),
    ("assistant", "Berlin."),
    ("user",      "One more — what was the movie you recommended earlier?"),
    ("user",      "Do you remember? It was at the start of our conversation."),
]


# ---------------------------------------------------------------------------
# TODO 1: Implement summarise_turns.
#         Given a list of ChatMessage objects, ask the LLM to compress them.
#         - Flatten `turns` into a single transcript string ("<role>: <content>" per line).
#         - Build a prompt asking the model to summarise the exchange in 2–3 sentences,
#           keeping key facts / decisions / topics and dropping conversational filler,
#           followed by the transcript.
#         - Send it as one ChatMessage("user", ...) via `llm.chat(...)` with
#           ChatOptions(temperature=0) so the summary is deterministic.
#         - Return a single ChatMessage with role "system" whose content begins with
#           "Summary of earlier conversation: " and then the model's text (this prefix
#           is what maybe_compact and the model rely on to recognise the summary).
# ---------------------------------------------------------------------------
def summarise_turns(turns: list[ChatMessage], llm) -> ChatMessage:
    raise NotImplementedError("TODO: implement summarise_turns")


# ---------------------------------------------------------------------------
# TODO 2: Implement maybe_compact.
#         1. Count the total tokens in `messages`.
#         2. If the count exceeds `budget`:
#            a. Separate system messages from non-system messages.
#            b. Identify the oldest non-system turns to summarise
#               (enough to bring the total below budget).
#            c. Call summarise_turns on those oldest turns.
#            d. Return: [system_messages] + [summary_message] + [remaining_turns].
#         3. If within budget, return messages unchanged.
#         4. Print a line when compaction fires so you can observe it.
# ---------------------------------------------------------------------------
def maybe_compact(
    messages: list[ChatMessage],
    budget: int,
    llm,
) -> list[ChatMessage]:
    # - Use count_history_tokens(messages); if it is within `budget`, return messages
    #   unchanged (fast path — no LLM call).
    # - Otherwise print a line noting compaction fired (so you can watch it in the demo),
    #   then partition messages into system messages and conversation (non-system) turns.
    # - Pick how many of the OLDEST conversation turns to compress — summarising roughly
    #   the first half (at least one) is a reasonable split. Call summarise_turns on that
    #   oldest slice and keep the rest verbatim.
    # - Reassemble in order: system messages, then the single summary message, then the
    #   kept recent turns — and return that new list.
    raise NotImplementedError("TODO: implement maybe_compact")


def main() -> None:
    print("=== Task 3: Conversation Memory / Compaction ===\n")
    print(f"Token budget: {TOKEN_BUDGET} tokens\n")

    try:
        llm = get_provider()
        print(f"Provider: {llm.name} / {llm.chat_model}\n")
    except Exception as e:
        print(f"Could not load provider: {e}")
        return

    # -------------------------------------------------------------------------
    # TODO 3: Simulate the conversation.
    #         For each turn in CONVERSATION_SCRIPT:
    #           1. Append the scripted message to history.
    #           2. If the turn is a "user" message, call maybe_compact, then
    #              call llm.chat(history) to get the assistant's reply.
    #           3. Append the assistant's reply to history.
    #           4. Print the current token count after each turn.
    # -------------------------------------------------------------------------

    history: list[ChatMessage] = [
        ChatMessage("system", "You are a helpful, knowledgeable assistant.")
    ]

    for role, content in CONVERSATION_SCRIPT:
        # Append the scripted message (both user and pre-written assistant turns).
        history.append(ChatMessage(role, content))  # type: ignore[arg-type]
        tokens = count_history_tokens(history)
        print(f"[{role:>9}] tokens={tokens:4d} | {content[:60]}...")

        if role == "user":
            try:
                # Compact if needed, then get a real LLM response.
                history = maybe_compact(history, TOKEN_BUDGET, llm)
                result = llm.chat(history)
                history.append(ChatMessage("assistant", result.text))
                print(f"[assistant] {result.text[:80]}...")
            except NotImplementedError:
                print("  [TODO: implement maybe_compact and summarise_turns]")
                # Append a placeholder so the loop can continue.
                history.append(ChatMessage("assistant", "(not implemented)"))
            except Exception as e:
                print(f"  ERROR: {e}")
                break

    print()
    print("Observation:")
    print("  When compaction fires, observe whether the model still answers questions")
    print("  about early turns (e.g. the movie recommendation) via the summary.")


if __name__ == "__main__":
    main()
